pkg/dataconn/wire.go

```diff
@@ -23,22 +28,27 @@ func NewWire(conn net.Conn) *Wire {
 }
 
 func (w *Wire) Write(msg *Message) error {
-	if err := binary.Write(w.writer, binary.LittleEndian, msg.MagicVersion); err != nil {
-		return err
-	}
-	if err := binary.Write(w.writer, binary.LittleEndian, msg.Seq); err != nil {
-		return err
-	}
-	if err := binary.Write(w.writer, binary.LittleEndian, msg.Type); err != nil {
-		return err
-	}
-	if err := binary.Write(w.writer, binary.LittleEndian, msg.Offset); err != nil {
-		return err
-	}
-	if err := binary.Write(w.writer, binary.LittleEndian, msg.Size); err != nil {
-		return err
-	}
-	if err := binary.Write(w.writer, binary.LittleEndian, uint32(len(msg.Data))); err != nil {
+	header := make([]byte, requestHeaderSize)
+	offset := 0
+
+	binary.LittleEndian.PutUint16(header[offset:], msg.MagicVersion)
+	offset += int(unsafe.Sizeof(msg.MagicVersion))
+
+	binary.LittleEndian.PutUint32(header[offset:], msg.Seq)
+	offset += int(unsafe.Sizeof(msg.Seq))
+
+	binary.LittleEndian.PutUint32(header[offset:], msg.Type)
+	offset += int(unsafe.Sizeof(msg.Type))
+
+	binary.LittleEndian.PutUint64(header[offset:], uint64(msg.Offset))
+	offset += int(unsafe.Sizeof(msg.Offset))
+
+	binary.LittleEndian.PutUint32(header[offset:], msg.Size)
+	offset += int(unsafe.Sizeof(msg.Size))
+
+	binary.LittleEndian.PutUint32(header[offset:], uint32(len(msg.Data)))
+
+	if _, err := w.writer.Write(header); err != nil {
```

Comment:
In general, it seems like we are trying to "buffer" the I/O to reduce system calls.  The `writer` is a `bufio.Writer` which seems buffer the I/O for us already.   Is the improvement because of less calls to `bufio.Writer.Write`?  I can't see how this could result in less system calls when the I/O is already using a buffer to reduce the amount of system calls?  Is the buffering not working?

Reply:
https://github.com/longhorn/longhorn-engine/pull/708#discussion_r899271299